knex/knex.d.ts

```diff
@@ -164,17 +164,30 @@ declare module "knex" {
 
     interface Join {
       (raw: Raw): QueryBuilder;
-      (tableName: string, callback: Function): QueryBuilder;
+      (tableName: string, callback: (joinClause: JoinClause) => any): QueryBuilder;
+      (tableName: string, columns: {[key: string]: string|Raw}): QueryBuilder;
+      (tableName: string, raw: Raw): QueryBuilder;
       (tableName: string, column1: string, column2: string): QueryBuilder;
       (tableName: string, column1: string, raw: Raw): QueryBuilder;
       (tableName: string, column1: string, operator: string, column2: string): QueryBuilder;
     }
 
+    interface JoinClause {
+      on(raw: Raw): JoinClause;
+      on(callback: Function): JoinClause;
```

Comment:
missing orOn, andOn, type and using: https://github.com/tgriesser/knex/blob/d40e578cadfc12c18ca21fbf436f64f9e36e6079/src/query/joinclause.js